monitoring/monitoring/monitoring.py

```diff
@@ -0,0 +1,219 @@
+        end = datetime.date(year, month, last_day_of_month) + datetime.timedelta(days=7)
+
+        date_format = '%Y-%m-%d'
+        start_str = datetime.date.strftime(start, date_format)
+        end_str = datetime.date.strftime(end, date_format)
+
+        invoice_month = datetime.date.strftime(start, '%Y%m')
+
+        cmd = f'''
+SELECT service.description as service_description, sku.description as sku_description, SUM(cost) as cost,
+CASE
+  WHEN EXISTS(SELECT 1 FROM UNNEST(labels) WHERE key = "namespace" and value = "default") THEN "batch-production"
+  WHEN EXISTS(SELECT 1 FROM UNNEST(labels) WHERE key = "namespace" and value LIKE '%pr-%') THEN "batch-test"
+  WHEN EXISTS(SELECT 1 FROM UNNEST(labels) WHERE key = "namespace") THEN "batch-dev"
+  WHEN EXISTS(SELECT 1 FROM UNNEST(labels) WHERE key = "role" and value LIKE 'vdc') THEN "k8s"
+  WHEN service.description = "Compute Engine" THEN "unknown"
+  ELSE NULL
+END AS source
+FROM `broad-ctsa.hail_billing.gcp_billing_export_v1_0055E5_9CA197_B9B894`
+WHERE DATE(_PARTITIONTIME) >= "{start_str}" AND DATE(_PARTITIONTIME) <= "{end_str}" AND project.name = "hail-vdc" AND invoice.month = "{invoice_month}"
+GROUP BY service_description, sku_description, source;
+'''
+
+        records = [x async for x in await bigquery_client.query(cmd)]
+
+        data = [(year, month, record['service_description'], record['sku_description'], record['source'], record['cost'])
+                for record in records]
+
+        @transaction(db)
+        async def insert(tx):
+            await tx.just_execute('''
+DELETE FROM monitoring_billing_data WHERE year = %s AND month = %s;
+''',
+                                  (year, month))
+
+            await tx.execute_many('''
+INSERT INTO monitoring_billing_data (year, month, service_description, sku_description, source, cost)
+VALUES (%s, %s, %s, %s, %s, %s);
+''',
+                                  data)
+
+        await insert()  # pylint: disable=no-value-for-parameter
+
+    now = datetime.datetime.now()
+    await _query(now)
+    last_month = get_previous_month(now)
+    end_last_month = get_last_day_month(last_month)
+    if now < end_last_month + datetime.timedelta(days=7):
+        await _query(last_month)
+
+    now_msecs = time_msecs()
+    await db.execute_update('UPDATE monitoring_billing_mark SET mark = %s;', (now_msecs,))
+
+    should_wait = True
+    return should_wait
+
+
+async def polling_loop(app):
+    db = app['db']
+
+    while True:
+        try:
+            now = datetime.datetime.now()
+            row = await db.select_and_fetchone('SELECT mark FROM monitoring_billing_mark;')
+            if not row['mark'] or now > datetime.datetime.fromtimestamp(row['mark'] / 1000) + datetime.timedelta(days=1):
+                app['query_billing_event'].set()
+                log.info('set event')
+        except Exception:
+            log.exception('error while polling for billing mark', exc_info=True)
```

Comment:
I think you can ditch the `try-except` and use `retry_long_running` (but without `run_if_changed_idempotent`).

Reply:
e.g. `bump_loop` in batch's scheduler